### PR TITLE
Add mechanism for early http header validation

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently we only check the http header objections after the full request has
been read. This commit offers the ability to validate the headers prior to
reading the full message and short circuit as appropriate.

### CODE REVIEW

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+    private void validationFailure(ChannelHandlerContext ctx, HeaderValidationException e) {
+        assert ctx.channel().eventLoop().inEventLoop();
+        state = e.shouldCloseChannel() ? STATE.DROPPING_DATA_PERMANENTLY : STATE.HANDLING_QUEUED_DATA;
+        HttpMessage messageToForward = (HttpMessage) pending.remove();
+        boolean fullRequestConsumed;
+        if (messageToForward instanceof LastHttpContent toRelease) {
+            messageToForward = (HttpMessage) toRelease.replace(Unpooled.EMPTY_BUFFER);
+            toRelease.release();
+            fullRequestConsumed = true;
```
Comment (albertzaharovits) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java:154]: I don't understand the reason for this, can you please explain why we need to clear the buffer only if the first message is also the last one? What's special about a http request that only has a `LastHttpContent` message.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+            pending.add(ReferenceCountUtil.retain(httpObject));
+            // Immediately return as this can only happen from a reentrant read(). We do not want to change
+            // autoread in this case.
+            return;
+        } else if (state == STATE.FORWARDING_DATA) {
+            assert pending.isEmpty();
+            if (httpObject instanceof LastHttpContent) {
+                state = STATE.WAITING_TO_START;
+            }
+            ctx.fireChannelRead(httpObject);
```
Comment (albertzaharovits) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java:63]: #### Background

Security is used to convey the authentication information (when successful) in thread-local variables that we call the `threadContext`.
In essence, all the code that's run by a thread with some authentication information in the `threadContext` is said to run "as" the "subject" from that authentication information.
Today, the `SecurityRestFilter` sets and clears the `threadContext`, in a try-block, before and after executing the request handler.

#### Question

Assuming authentication is successful, and here we're forwarding `httpObject`s to the next pipeline handler. We need to ensure that the `threadContext` contains the authentication information, when the **complete request** gets to the `RestController`.

Is is correct to set the `threadContext` to contain the authentication when we forward the `LastHttpContent` message?

I think the downstream pipeline handler, upon seeing this message will assemble the complete request and forward it further, **under the same thread call stack**, i.e. not enqueue it, like it does with the other `httpObject`s and wait for some other event to forward the complete request, in which case the `threadContext` will not be populated correctly.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java (hunk 1 above)
Comment (Tim-Brooks) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java:154]: If the first message (headers) is also the last "content", it is possible that actually byte buffer content is included in the message. We clear that here before sending the exception down the pipeline.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java (hunk 2 above)
Comment (Tim-Brooks) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java:63]: Yeah. We cannot propagate this through thread locals at this point. As additional context switching could happen. If there is information from Authentication that we need to propagate, the Validator will have to return it and we can find some mechanism to shove it in the http request object until "handling" where we can set the thread context.

Comment (albertzaharovits) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java:63]: > we can find some mechanism to shove it in the http request object until "handling" where we can set the thread context.

We discussed to wrap and decorate the `HttpRequest` that we forward with the authn result. That looks good, but the problem is that the `HttpRequest` is not available from the `FullHttpRequest` that the `RestController` ultimately deals with, so it cannot get hold of the authn result either (in order to set it in the thread context that runs the request handler).
In order to expose the authn result to the `FullHttpRequest` we'll need to extend the `HttpObjectAggregator` slightly (more precisely extend `#beginAggregation` to return an implementation of `FullHttpMessage` that contains the authn result from the `HttpMessage start` parameter). 

I'm going to try implementing this.

**Comment by albertzaharovits** [comment]
@Tim-Brooks I'm going to push to this branch of yours in order to get it merged.
The intent is for it to work to support the strategy described in https://github.com/elastic/elasticsearch/pull/93656 .

Let me know if you prefer I create a branch on my fork, incorporate the changes here, and iterate on that instead.

**Comment by albertzaharovits** [comment]
@Tim-Brooks This is ready for your review.

I'll be adding a few more tests soon, but I think this is now looking very polished.

**Comment by albertzaharovits** [comment]
@ping @Tim-Brooks

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java [hunk 3]
```diff
@@ -374,24 +382,24 @@ protected void initChannel(Channel ch) throws Exception {
...
+                @Override
+                protected boolean isContentAlwaysEmpty(HttpResponse msg) {
+                    // non-chunked responses (Netty4HttpResponse extends Netty's DefaultFullHttpResponse) with chunked transfer
+                    // encoding are only sent by us in response to HEAD requests and must always have an empty body
+                    if (msg instanceof Netty4HttpResponse netty4HttpResponse && HttpUtil.isTransferEncodingChunked(msg)) {
+                        assert netty4HttpResponse.content().isReadable() == false;
+                        return true;
                     }
-                })
-                .addLast("aggregator", aggregator);
```
Comment (Tim-Brooks) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java:394]: Can we remove this formatting change?

**Review by Tim-Brooks — COMMENTED** [review]
One minor request

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java (hunk 3 above)
Comment (albertzaharovits) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java]: I've pushed [d6d7b87](https://github.com/elastic/elasticsearch/pull/92220/commits/d6d7b8739eca663d094fdfc8e22b9cccc638c187) to stuff some inline comments following an `addLast` such that spotless doesn't kick in to combine multiple `addLast` on the same line.

**Comment by Tim-Brooks** [comment]
LGTM

**Review by jakelandis — APPROVED** [review]
Note - I only have a surface level understanding of the change. However since this was join effort by Tim and Albert I wanted to provided a 3rd party approval.

**Review by albertzaharovits — APPROVED** [review]
LGTM 💯

**Pull request merged**